crates/wasi-http/src/p3/response.rs

```diff
@@ -70,17 +71,26 @@ impl Response {
                 // `Content-Length` header value is validated in `fields` implementation
                 let content_length =
                     get_content_length(&res.headers).context("failed to parse `content-length`")?;
-                GuestBody::new(
-                    store,
-                    contents_rx,
-                    trailers_rx,
-                    result_tx,
-                    fut,
-                    content_length,
-                    ErrorCode::HttpResponseBodySize,
-                    getter,
-                )
-                .boxed()
+                let guest_body = move |store, contents_rx| {
+                    GuestBody::new(
+                        store,
+                        contents_rx,
+                        trailers_rx,
+                        result_tx,
+                        fut,
+                        content_length,
+                        ErrorCode::HttpResponseBodySize,
+                        getter,
+                    )
+                    .boxed()
+                };
+                match contents_rx
+                    .map(|rx| rx.try_into::<HostBodyStreamProducer<T>>(store.as_context_mut()))
+                {
+                    Some(Ok(mut producer)) => mem::take(&mut producer.body),
+                    Some(Err(rx)) => guest_body(store, Some(rx)),
+                    None => guest_body(store, None),
+                }
```

Comment:
Could this happen during the construction of `Response` instead of here? Then I believe this would naturally fall through to `Body::Host` below?

Reply:
Good idea; done!